### PR TITLE
[advanced reboot] Add Paramiko module for device connection

### PR DESCRIPTION
**Pull request by tahmed-dev**
### Description of PR
Parmiko module provides fallback mechanism to using username/password
This is required if we are rebooting into new image using advanced
reboot test fixture.

signed-off-by: Tamer Ahmed <tamer.ahmed@microsoft.com>

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### CODE REVIEW

File: ansible/roles/test/files/ptftests/device_connection.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import logging
+from paramiko.ssh_exception import BadHostKeyException, AuthenticationException, SSHException
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_CMD_EXECUTION_TIMEOUT_SEC = 10
+
+class DeviceConnection:
+    '''
+    DeviceConnection uses Pramiko module to connect to devices
```
Comment (neethajohn) [inline ansible/roles/test/files/ptftests/device_connection.py:11]: spelling check

File: ansible/roles/test/files/ptftests/device_connection.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    Pramiko module uses fallback mechanism where it would first try to use
+    ssh key and that fails, it will attempt username/password combination
```
Comment (neethajohn) [inline ansible/roles/test/files/ptftests/device_connection.py:14]: 'paramiko' spell check

Comment (tahmed-dev) [reply ansible/roles/test/files/ptftests/device_connection.py]: Thanks. done

File: tests/common/fixtures/advanced_reboot.py [hunk 3]
```diff
@@ -115,9 +115,14 @@ def __buildTestbedData(self):
 
         self.rebootData['dut_hostname'] = self.mgFacts['minigraph_mgmt_interface']['addr']
         self.rebootData['dut_mac'] = hostFacts['ansible_Ethernet0']['macaddress']
-        self.rebootData['dut_username'] = hostFacts['ansible_env']['SUDO_USER']
         self.rebootData['vlan_ip_range'] = self.mgFacts['minigraph_vlan_interfaces'][0]['subnet']
         self.rebootData['dut_vlan_ip'] = self.mgFacts['minigraph_vlan_interfaces'][0]['addr']
+
+        invetory = self.duthost.host.options['inventory'].split('/')[-1]
+        secrets = self.duthost.host.options['variable_manager']._hostvars[self.duthost.hostname]['secret_group_vars']
```
Comment (msosyak) [inline tests/common/fixtures/advanced_reboot.py:122]: I faced `KeyError: ('secret_group_vars',) ` when tried to run `platform/test_advanced_reboot.py::test_fast_reboot` which uses this fixture.

In the next two lines, you tried to get `sonicadmin_user` and `sonicadmin_password`  from `secrets`
I have `sonicadmin_user` and `sonicadmin_password` directly in `self.duthost.host.options['variable_manager']._hostvars[self.duthost.hostname]`

The values are defined in  `ansible/group_vars/lab/secrets.yml`.

I, Maybe, missed something in testbed definition, but do we really need to get  `sonicadmin_user` and `sonicadmin_password` from `secret_group_vars`? Maybe there is a mistake? 

If No, then where secret_group_vars should be defined?

@tahmed-dev  Could you please help me?

Comment (tahmed-dev) [reply tests/common/fixtures/advanced_reboot.py:122]: @msosyak Thanks for your question.

Those secrets are defined in secrets.json file (ansible/group_vars/all/secrets.json). Here is the format of the file:

`{
    .
    .
    "secret_group_vars": {
        "str": {
            "ansible_ssh_pass": "",
            "ansible_become_pass": "",
            "sonicadmin_user": "",
            "sonicadmin_password": "",
            "sonicadmin_password": "",
            "sonicadmin_initial_password": "",
            "lab_admin_pass": "",
            "ptf_host_user": "",
            "ptf_host_pass": "",
            "fanout_admin_user": "",
            "fanout_admin_password": ""
        },
    },
    .
    .
}
`
Where str in the above file is the name of the inventory file. You may wanna change it the name of the inventory file used in your setup.

**Pull request merged**